barretenberg/cpp/src/barretenberg/commitment_schemes/commitment_key.hpp

```diff
@@ -173,6 +175,147 @@ template <class Curve> class CommitmentKey {
+        }
+
+        // Compute "active" percentage of polynomial; resort to standard commit if appropriate
+        size_t percentage_nonzero = total_num_scalars * 100 / polynomial.size();
+        if (percentage_nonzero > NONZERO_THRESHOLD) {
+            return commit(polynomial);
+        }
+
+        // Extract the precomputed point table (contains raw SRS points at even indices and the corresponding
+        // endomorphism point (\beta*x, -y) at odd indices).
+        std::span<G1> point_table = srs->get_monomial_points();
+
+        std::vector<Fr> scalars;
+        scalars.reserve(total_num_scalars);
+        for (const auto& range : active_ranges) {
+            auto start = &polynomial[range.first];
+            auto end = &polynomial[range.second];
+            scalars.insert(scalars.end(), start, end);
+        }
+        std::vector<G1> points;
+        points.reserve(total_num_scalars * 2);
+        for (const auto& range : active_ranges) {
+            auto start = &point_table[2 * range.first];
+            auto end = &point_table[2 * range.second];
+            points.insert(points.end(), start, end);
+        }
+
+        // Call pippenger
+        return scalar_multiplication::pippenger_unsafe<Curve>(scalars, points, pippenger_runtime_state);
+    }
+
+    /**
+     * @brief Efficiently commit to a polynomial with discrete blocks of arbitrary elements and constant elements
+     * @details Similar to method commit_structured() except the complement to the "active" region cantains non-zero
+     * constant values (which are assumed to differ between blocks). This is exactly the structure of the permutation
+     * grand product polynomial z_perm when a structured execution trace is in use.
+     * @warning Requires a copy of all {point, scalar} pairs (including endo points) corresponding to the primary blocks
+     * and a copy of all of the points (without endo points) corresponding to their complement.
+     *
+     * @param polynomial
+     * @param active_ranges
+     * @return Commitment
+     */
+    Commitment commit_structured_with_nonzero_complement(PolynomialSpan<const Fr> polynomial,
+                                                         const std::vector<std::pair<size_t, size_t>>& active_ranges)
+    {
+        BB_OP_COUNT_TIME();
+        ASSERT(polynomial.end_index() <= srs->get_monomial_size());
+
+        using BatchedAddition = BatchedAffineAddition<Curve>;
+
+        // Percentage of constant coefficients below which we resort to the conventional commit method
+        const size_t CONSTANT_THRESHOLD = 50;
+
+        // Compute the active range complement over which the polynomial is assumed to be constant within each range
+        std::vector<std::pair<size_t, size_t>> active_ranges_complement;
+        for (size_t i = 0; i < active_ranges.size(); ++i) {
+            size_t start = active_ranges[i].second;
+            size_t end = active_ranges[i + 1].first;
```

Comment:
doesn't `i+1` go out of bounds here?

Reply:
It does indeed, thanks. It was working correctly because the bad value gets immediately set to the correct thing on the line below but I've updated the loop to not overindex